### ISSUE TITLE
Exposing generators

### ISSUE THREAD

**Issue by HDembinski** [dsaxton/resample#27]
I have been thinking about the common interface for bootstrap and jackknife. I think in addition to the functions which compute the replicas, `jackknife` and `bootstrap`, we should expose the generators for these samples, before they are passed through the statistical function. You already had this before, if the function argument was `None`. The problem with that old approach was that the whole collection of resampled samples can be become too large to fit into memory. But we can avoid this or at least reduce this with generators. The generator computes only one sample at a time. In most cases, the state of the generator is much smaller than the state of the resampled samples all held in memory at once.

So, long story short, I think it would be great to
- fully expose the currently hidden jackknife generator
- refactor the bootstrapping code inside `bootstrap` into a generator
- make the old `bootstrap` function use the generator

I would like to work on this.

**Comment by dsaxton** [comment]
Good ideas, I also think the behavior of returning raw bootstrap samples should be removed with the bootstrap function _requiring_ a function and always returning the output of that function. It would still be possible to blow up memory by returning some high dimensional output, but at least it's then discouraged.

I think this came up before but would also need to think about how to implement the balanced bootstrap using generators. I am not sure if there is a lazy way to do this, but the way it works now is essentially to take an array [0, 1, 2], repeat it b times [0, 1, 2, 0, 1, 2, 0, 1, 2], take a permutation, and then repartition it into samples. Maybe in numba a simple algorithm using decrementing counts and a loop could be used instead?

**Comment by HDembinski** [comment]
I fully agree to your first point.

I kept thinking about it, but I cannot imagine a solution that does not hold at least an array of indices of size N x B in memory, where N is the sample size and B the number or bootstrap replicas. However, this array can use a suitably small integer type, like a short, so there are some savings if the sample type is double, and even more when the input is a vector per observation.

**Comment by HDembinski** [comment]
I started to work on that in #28

**Comment by dsaxton** [comment]
> I started to work on that in #28

Thanks!
